// @progress/kendo-angular-dateinputs/dist/es/timepicker/localization/timepicker-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
var TimePickerCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TimePickerCustomMessagesComponent, _super);
    function TimePickerCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    TimePickerCustomMessagesComponent_1 = TimePickerCustomMessagesComponent;
    Object.defineProperty(TimePickerCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    TimePickerCustomMessagesComponent = TimePickerCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: Messages,
                    useExisting: forwardRef(function () { return TimePickerCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                }
            ],
            selector: 'kendo-timepicker-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], TimePickerCustomMessagesComponent);
    return TimePickerCustomMessagesComponent;
    var TimePickerCustomMessagesComponent_1;
}(Messages));
export { TimePickerCustomMessagesComponent };
