// @progress/kendo-charts/dist/es/common/constants.js
var X = "x";
var Y = "y";
var TOP = "top";
var BOTTOM = "bottom";
var LEFT = "left";
var RIGHT = "right";
var CENTER = "center";
var WIDTH = "width";
var HEIGHT = "height";
var COORD_PRECISION = 3;
var MAX_VALUE = Number.MAX_VALUE;
var MIN_VALUE = -Number.MAX_VALUE;
var DEFAULT_WIDTH = 600;
var DEFAULT_HEIGHT = 400;
var WHITE = "#fff";
var BLACK = "#000";
var DEFAULT_FONT = "12px sans-serif";
var DEFAULT_PRECISION = 10;
var AXIS_LABEL_CLICK = "axisLabelClick";
var NOTE_CLICK = "noteClick";
var NOTE_HOVER = "noteHover";
var OUTSIDE = "outside";
var NONE = "none";
var CIRCLE = "circle";
var TRIANGLE = "triangle";
var CROSS = "cross";
var ARC = "arc";
var INSIDE = "inside";
var VALUE = "value";
var STRING = "string";
var OBJECT = "object";
var DATE = "date";
var FORMAT_REGEX = /\{\d+:?/;
var HIGHLIGHT_ZINDEX = 100;

export {
    X, Y, WIDTH, HEIGHT, DEFAULT_HEIGHT, DEFAULT_WIDTH,
    TOP, LEFT, BOTTOM, RIGHT, CENTER,
    COORD_PRECISION, DEFAULT_PRECISION,
    CIRCLE, TRIANGLE, CROSS,
    MAX_VALUE, MIN_VALUE,
    WHITE, BLACK, DEFAULT_FONT,
    AXIS_LABEL_CLICK, OUTSIDE, INSIDE,
    NONE, NOTE_CLICK,
    NOTE_HOVER, VALUE, STRING, OBJECT,
    DATE, ARC, FORMAT_REGEX, HIGHLIGHT_ZINDEX
};
