// @progress/kendo-angular-dateinputs/dist/es2015/calendar/navigation.component.js
import * as tslib_1 from "tslib";
/* tslint:disable:component-selector-name  component-selector-type */
import { Component, ChangeDetectionStrategy, EventEmitter, HostBinding, Input, Output, ViewChild } from '@angular/core';
import { BusViewService } from './services/bus-view.service';
import { DOMService } from './services/dom.service';
import { CalendarViewEnum } from './models/view.enum';
import { VirtualizationComponent } from '../virtualization/virtualization.component';
import { MIN_DATE, MAX_DATE } from '../defaults';
import { cloneDate } from '@progress/kendo-date-math';
import { dateInRange } from '../util';
const ITEMS_COUNT = 30;
/**
 * @hidden
 */
let NavigationComponent = class NavigationComponent {
    constructor(bus, dom) {
        this.bus = bus;
        this.dom = dom;
        this.min = new Date(MIN_DATE);
        this.max = new Date(MAX_DATE);
        this.focusedDate = new Date();
        this.valueChange = new EventEmitter();
        this.dates = [];
        this.take = ITEMS_COUNT;
        this.indexToScroll = -1;
    }
    get getComponentClass() {
        return true;
    }
    ngOnInit() {
        this.dom.ensureHeights();
        const calendarHeight = this.dom.calendarHeight;
        this.itemHeight = this.dom.navigationItemHeight;
        this.maxViewHeight = this.dom.monthViewHeight;
        this.topOffset = (calendarHeight - this.itemHeight) / 2;
        this.bottomOffset = calendarHeight - this.itemHeight;
    }
    ngOnChanges(changes) {
        this.service = this.bus.service(this.activeView);
        if (!this.service) {
            return;
        }
        const viewDate = dateInRange(this.focusedDate, this.min, this.max);
        const total = this.service.total(this.min, this.max);
        const totalChanged = this.total && this.total !== total;
        this.skip = this.service.skip(viewDate, this.min);
        this.total = total;
        if (totalChanged || !this.service.isInArray(viewDate, this.dates)) {
            this.dates = this.service.datesList(viewDate, this.getTake(this.skip));
        }
        if (!!changes.focusedDate || totalChanged) {
            this.indexToScroll = this.service.skip(this.focusedDate, this.min);
        }
    }
    ngAfterViewInit() {
        if (this.indexToScroll === -1) {
            return;
        }
        this.virtualization.scrollToIndex(this.indexToScroll);
        this.indexToScroll = -1;
    }
    ngAfterViewChecked() {
        if (this.indexToScroll === -1) {
            return;
        }
        this.virtualization.scrollToIndex(this.indexToScroll);
        this.indexToScroll = -1;
    }
    pageChange({ skip }) {
        this.dates = this.service.datesList(this.service.addToDate(this.min, skip), this.getTake(skip));
    }
    scrollChange({ offset }) {
        const translate = `translateY(${offset}px)`;
        this.style = { transform: translate, '-ms-transform': translate };
    }
    handleDateChange(candidate) {
        this.valueChange.emit(cloneDate(candidate));
    }
    getTake(skip) {
        return Math.min(this.total - skip, this.take);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], NavigationComponent.prototype, "activeView", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Date)
], NavigationComponent.prototype, "min", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Date)
], NavigationComponent.prototype, "max", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Date)
], NavigationComponent.prototype, "focusedDate", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], NavigationComponent.prototype, "valueChange", void 0);
tslib_1.__decorate([
    ViewChild(VirtualizationComponent),
    tslib_1.__metadata("design:type", VirtualizationComponent)
], NavigationComponent.prototype, "virtualization", void 0);
tslib_1.__decorate([
    HostBinding("class.k-calendar-navigation"),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], NavigationComponent.prototype, "getComponentClass", null);
NavigationComponent = tslib_1.__decorate([
    Component({
        changeDetection: ChangeDetectionStrategy.OnPush,
        selector: 'kendo-calendar-navigation',
        template: `
    <span class="k-calendar-navigation-highlight"></span>
    <kendo-virtualization
        [skip]="skip"
        [take]="take"
        [total]="total"
        [itemHeight]="itemHeight"
        [topOffset]="topOffset"
        [bottomOffset]="bottomOffset"
        [maxScrollDifference]="maxViewHeight"
        (pageChange)="pageChange($event)"
        (scrollChange)="scrollChange($event)"
    >
        <ul [ngStyle]="style" class="k-reset">
            <li *kFor="let date of dates" (click)="handleDateChange(date)">
                <span [class.k-calendar-navigation-marker]="service.isRangeStart(date)">
                    {{service.navigationTitle(date)}}
                </span>
            </li>
        </ul>
    </kendo-virtualization>
  `
    }),
    tslib_1.__metadata("design:paramtypes", [BusViewService, DOMService])
], NavigationComponent);
export { NavigationComponent };
