// @progress/kendo-angular-layout/dist/es/splitter/splitter-bar.component.js
import * as tslib_1 from "tslib";
import { Component, Host, HostBinding, HostListener, Input, Optional, Inject, NgZone } from '@angular/core';
import { RTL } from '@progress/kendo-angular-l10n';
import { DraggableDirective } from '../common/draggable.directive';
import { Keys } from '../common/keys';
import { SplitterService } from './splitter.service';
import { Subscription } from 'rxjs/Subscription';
import { of } from 'rxjs/observable/of';
import { delay } from 'rxjs/operators/delay';
import { filter } from 'rxjs/operators/filter';
import { map } from 'rxjs/operators/map';
import { switchMap } from 'rxjs/operators/switchMap';
import { takeUntil } from 'rxjs/operators/takeUntil';
import { tap } from 'rxjs/operators/tap';
var stopPropagation = function (_a) {
    var event = _a.originalEvent;
    event.stopPropagation();
    event.preventDefault();
};
var toggleContentOverlay = function (splitter, index, toggle) {
    splitter.pane(index).overlayContent = toggle;
    splitter.pane(index + 1).overlayContent = toggle;
};
var preventOnDblClick = function (release) { return function (mouseDown) {
    return of(mouseDown).pipe(delay(150), takeUntil(release));
}; };
var classFromObject = function (classes) { return Object.keys(classes).filter(function (c) { return classes[c]; }).join(' '); };
var createMoveStream = function (draggable) { return function (mouseDown) {
    return draggable.kendo.drag
        .pipe(takeUntil(draggable.kendo.release), map(function (_a) {
        var pageX = _a.pageX, pageY = _a.pageY;
        return ({
            originalX: mouseDown.pageX,
            originalY: mouseDown.pageY,
            pageX: pageX,
            pageY: pageY
        });
    }));
}; };
/**
 * @hidden
 */
var SplitterBarComponent = /** @class */ (function () {
    function SplitterBarComponent(draggable, splitter, zone, rtl) {
        this.draggable = draggable;
        this.splitter = splitter;
        this.zone = zone;
        this.orientation = 'horizontal';
        this.index = 0;
        this.ariaRole = 'separator';
        this.focused = false;
        this.subscriptions = new Subscription();
        this.direction = rtl ? 'rtl' : 'ltr';
    }
    Object.defineProperty(SplitterBarComponent.prototype, "tabIndex", {
        get: function () {
            return this.splitter.isStatic(this.index) ? -1 : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplitterBarComponent.prototype, "hostClasses", {
        get: function () {
            var isHorizontal = this.orientation === 'horizontal';
            var isDraggable = this.splitter.isDraggable(this.index);
            var isStatic = this.splitter.isStatic(this.index);
            return classFromObject({
                'k-state-focused': this.focused,
                'k-splitbar': true,
                'k-splitbar-horizontal': isHorizontal,
                'k-splitbar-vertical': !isHorizontal,
                'k-splitbar-draggable-horizontal': isHorizontal && isDraggable,
                'k-splitbar-draggable-vertical': !isHorizontal && isDraggable,
                'k-splitbar-static-horizontal': isHorizontal && isStatic,
                'k-splitbar-static-vertical': !isHorizontal && isStatic
            });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplitterBarComponent.prototype, "order", {
        get: function () {
            return 2 * this.index + 1;
        },
        enumerable: true,
        configurable: true
    });
    SplitterBarComponent.prototype.collapseAny = function () {
        this.tryToggleNearest();
    };
    SplitterBarComponent.prototype.onFocusIn = function () {
        this.focused = true;
    };
    SplitterBarComponent.prototype.onFocusOut = function () {
        this.focused = false;
    };
    SplitterBarComponent.prototype.onKeyDown = function (event) {
        var _this = this;
        var keyCode = event && event.keyCode;
        var isHorizontal = this.orientation === 'horizontal';
        var resize = function (delta) {
            event.preventDefault();
            var state = _this.splitter.dragState(_this.index);
            // resizing next pane, reverse delta
            if (state.index === _this.index + 1) {
                delta *= -1;
            }
            _this.splitter.setSize(state, delta);
        };
        if (keyCode === Keys.enter) {
            event.preventDefault();
            this.tryToggleNearest();
        }
        else if (isHorizontal && keyCode === Keys.left) {
            resize(-10);
        }
        else if (isHorizontal && keyCode === Keys.right) {
            resize(10);
        }
        else if (!isHorizontal && keyCode === Keys.up) {
            resize(-10);
        }
        else if (!isHorizontal && keyCode === Keys.down) {
            resize(10);
        }
    };
    SplitterBarComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscriptions.add(this.zone.runOutsideAngular(function () {
            var state;
            _this.draggable.kendo.press.pipe(tap(stopPropagation), filter(function () { return _this.splitter.isDraggable(_this.index); }), tap(function () { return state = _this.splitter.dragState(_this.index); }), tap(function () { return toggleContentOverlay(_this.splitter, _this.index, true); }), switchMap(preventOnDblClick(_this.draggable.kendo.release)), switchMap(createMoveStream(_this.draggable))).subscribe(function (_a) {
                var pageX = _a.pageX, pageY = _a.pageY, originalX = _a.originalX, originalY = _a.originalY;
                var delta;
                if (_this.orientation === 'vertical') {
                    delta = pageY - originalY;
                }
                else if (_this.direction === 'rtl') {
                    delta = originalX - pageX;
                }
                else {
                    delta = pageX - originalX;
                }
                // resizing next pane, reverse delta
                if (state.index === _this.index + 1) {
                    delta *= -1;
                }
                _this.splitter.setSize(state, delta);
            });
        }));
        this.subscriptions.add(this.draggable.kendo.release.subscribe(function () { return toggleContentOverlay(_this.splitter, _this.index, false); }));
    };
    SplitterBarComponent.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            this.subscriptions.unsubscribe();
        }
    };
    SplitterBarComponent.prototype.togglePrevious = function () {
        this.splitter.tryToggle(this.index);
    };
    SplitterBarComponent.prototype.toggleNext = function () {
        this.splitter.tryToggle(this.index + 1);
    };
    SplitterBarComponent.prototype.previousArrowClass = function () {
        var pane = this.splitter.pane(this.index);
        var isCollapsible = pane.collapsible;
        var isCollapsed = pane.collapsed;
        var isHorizontal = this.orientation === 'horizontal';
        return classFromObject({
            'k-icon': true,
            'k-hidden': !isCollapsible,
            'k-collapse-prev': isCollapsible,
            'k-i-arrow-60-left': isCollapsible && isHorizontal && !isCollapsed,
            'k-i-arrow-60-right': isCollapsible && isHorizontal && isCollapsed,
            'k-i-arrow-60-up': isCollapsible && !isHorizontal && !isCollapsed,
            'k-i-arrow-60-down': isCollapsible && !isHorizontal && isCollapsed
        });
    };
    SplitterBarComponent.prototype.nextArrowClass = function () {
        var pane = this.splitter.pane(this.index + 1);
        var isCollapsible = pane.collapsible;
        var isCollapsed = pane.collapsed;
        var isHorizontal = this.orientation === 'horizontal';
        return classFromObject({
            'k-icon': true,
            'k-hidden': !isCollapsible,
            'k-collapse-next': isCollapsible,
            'k-i-arrow-60-right': isCollapsible && isHorizontal && !isCollapsed,
            'k-i-arrow-60-left': isCollapsible && isHorizontal && isCollapsed,
            'k-i-arrow-60-down': isCollapsible && !isHorizontal && !isCollapsed,
            'k-i-arrow-60-up': isCollapsible && !isHorizontal && isCollapsed
        });
    };
    SplitterBarComponent.prototype.tryToggleNearest = function () {
        var prev = this.index;
        var next = this.index + 1;
        if (!this.splitter.tryToggle(prev)) {
            this.splitter.tryToggle(next);
        }
    };
    tslib_1.__decorate([
        Input(),
        HostBinding('attr.aria-orientation'),
        tslib_1.__metadata("design:type", String)
    ], SplitterBarComponent.prototype, "orientation", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], SplitterBarComponent.prototype, "index", void 0);
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", String)
    ], SplitterBarComponent.prototype, "ariaRole", void 0);
    tslib_1.__decorate([
        HostBinding('attr.tabindex'),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [])
    ], SplitterBarComponent.prototype, "tabIndex", null);
    tslib_1.__decorate([
        HostBinding('class'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [])
    ], SplitterBarComponent.prototype, "hostClasses", null);
    tslib_1.__decorate([
        HostBinding('class.k-state-focused'),
        tslib_1.__metadata("design:type", Boolean)
    ], SplitterBarComponent.prototype, "focused", void 0);
    tslib_1.__decorate([
        HostBinding('style.order'),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [])
    ], SplitterBarComponent.prototype, "order", null);
    tslib_1.__decorate([
        HostListener('dblclick'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], SplitterBarComponent.prototype, "collapseAny", null);
    tslib_1.__decorate([
        HostListener('focusin'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], SplitterBarComponent.prototype, "onFocusIn", null);
    tslib_1.__decorate([
        HostListener('focusout'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], SplitterBarComponent.prototype, "onFocusOut", null);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], SplitterBarComponent.prototype, "onKeyDown", null);
    SplitterBarComponent = tslib_1.__decorate([
        Component({
            selector: 'kendo-splitter-bar',
            template: "\n      <div [class]=\"previousArrowClass()\" (click)=\"togglePrevious()\"></div>\n      <div class=\"k-resize-handle\"></div>\n      <div [class]=\"nextArrowClass()\" (click)=\"toggleNext()\"></div>\n    "
        }),
        tslib_1.__param(0, Host()),
        tslib_1.__param(3, Optional()), tslib_1.__param(3, Inject(RTL)),
        tslib_1.__metadata("design:paramtypes", [DraggableDirective,
            SplitterService,
            NgZone, Boolean])
    ], SplitterBarComponent);
    return SplitterBarComponent;
}());
export { SplitterBarComponent };
