// @progress/kendo-angular-layout/dist/es2015/splitter/splitter-pane.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, HostBinding, Input, Output } from '@angular/core';
/**
 * Represents the Kendo UI Splitter Pane component for Angular.
 */
let SplitterPaneComponent = class SplitterPaneComponent {
    constructor(element) {
        this.element = element;
        /**
         * Specifies if the user is allowed to resize the pane and provide space for other panes.
         */
        this.resizable = true;
        /**
         * Specifies if the user is allowed to hide the pane and provide space for other panes.
         */
        this.collapsible = false;
        /**
         * Specifies if overflowing content is scrollable or hidden.
         */
        this.scrollable = true;
        /**
         * Specifies if the pane is initially collapsed.
         */
        this.collapsed = false;
        /**
         * @hidden
         */
        this.orientation = 'horizontal';
        /**
         * @hidden
         */
        this.containsSplitter = false;
        /**
         * @hidden
         */
        this.overlayContent = false;
        /**
         * Fires each time the user resizes the Splitter pane.
         * The event data contains the new pane size.
         * Allows a two-way binding of the pane `size` property.
         */
        this.sizeChange = new EventEmitter();
        /**
         * Fires each time the `collapsed` property changes.
         * The event data contains the new property state.
         * Allows a two-way binding of the `collapsed` pane property.
         */
        this.collapsedChange = new EventEmitter();
        this.hostClass = true;
    }
    get isHidden() {
        return this.collapsed;
    }
    get styleOrder() {
        return this.order;
    }
    get styleDisplayFlex() {
        return this.containsSplitter;
    }
    get styleHeight() {
        return this.size;
    }
    get staticPaneClass() {
        return !this.resizable && !this.collapsible || this.fixedSize;
    }
    get scrollablePaneClass() {
        return this.scrollable;
    }
    get fixedSize() {
        return this.size && this.size.length > 0;
    }
    /**
     * @hidden
     */
    get computedSize() {
        if (this.orientation === 'vertical') {
            return this.element.nativeElement.offsetHeight;
        }
        else {
            return this.element.nativeElement.offsetWidth;
        }
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], SplitterPaneComponent.prototype, "order", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], SplitterPaneComponent.prototype, "size", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], SplitterPaneComponent.prototype, "min", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], SplitterPaneComponent.prototype, "max", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "resizable", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "collapsible", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "scrollable", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "collapsed", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], SplitterPaneComponent.prototype, "orientation", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "containsSplitter", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "overlayContent", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], SplitterPaneComponent.prototype, "sizeChange", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], SplitterPaneComponent.prototype, "collapsedChange", void 0);
tslib_1.__decorate([
    HostBinding('hidden'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "isHidden", null);
tslib_1.__decorate([
    HostBinding('style.order'),
    tslib_1.__metadata("design:type", Number),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "styleOrder", null);
tslib_1.__decorate([
    HostBinding('class.k-pane-flex'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "styleDisplayFlex", null);
tslib_1.__decorate([
    HostBinding('style.flex-basis'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "styleHeight", null);
tslib_1.__decorate([
    HostBinding('class.k-pane'),
    tslib_1.__metadata("design:type", Boolean)
], SplitterPaneComponent.prototype, "hostClass", void 0);
tslib_1.__decorate([
    HostBinding('class.k-pane-static'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "staticPaneClass", null);
tslib_1.__decorate([
    HostBinding('class.k-scrollable'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SplitterPaneComponent.prototype, "scrollablePaneClass", null);
SplitterPaneComponent = tslib_1.__decorate([
    Component({
        exportAs: 'kendoSplitterPane',
        selector: 'kendo-splitter-pane',
        template: `
        <ng-container *ngIf="!collapsed"><ng-content></ng-content></ng-container>
        <div *ngIf="overlayContent" class="k-splitter-overlay k-overlay"></div>
    `
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef])
], SplitterPaneComponent);
export { SplitterPaneComponent };
