// @progress/kendo-charts/dist/es2015/common/constants.js
const X = "x";
const Y = "y";
const TOP = "top";
const BOTTOM = "bottom";
const LEFT = "left";
const RIGHT = "right";
const CENTER = "center";
const WIDTH = "width";
const HEIGHT = "height";
const COORD_PRECISION = 3;
const MAX_VALUE = Number.MAX_VALUE;
const MIN_VALUE = -Number.MAX_VALUE;
const DEFAULT_WIDTH = 600;
const DEFAULT_HEIGHT = 400;
const WHITE = "#fff";
const BLACK = "#000";
const DEFAULT_FONT = "12px sans-serif";
const DEFAULT_PRECISION = 10;
const AXIS_LABEL_CLICK = "axisLabelClick";
const NOTE_CLICK = "noteClick";
const NOTE_HOVER = "noteHover";
const OUTSIDE = "outside";
const NONE = "none";
const CIRCLE = "circle";
const TRIANGLE = "triangle";
const CROSS = "cross";
const ARC = "arc";
const INSIDE = "inside";
const VALUE = "value";
const STRING = "string";
const OBJECT = "object";
const DATE = "date";
const FORMAT_REGEX = /\{\d+:?/;
const HIGHLIGHT_ZINDEX = 100;

export {
    X, Y, WIDTH, HEIGHT, DEFAULT_HEIGHT, DEFAULT_WIDTH,
    TOP, LEFT, BOTTOM, RIGHT, CENTER,
    COORD_PRECISION, DEFAULT_PRECISION,
    CIRCLE, TRIANGLE, CROSS,
    MAX_VALUE, MIN_VALUE,
    WHITE, BLACK, DEFAULT_FONT,
    AXIS_LABEL_CLICK, OUTSIDE, INSIDE,
    NONE, NOTE_CLICK,
    NOTE_HOVER, VALUE, STRING, OBJECT,
    DATE, ARC, FORMAT_REGEX, HIGHLIGHT_ZINDEX
};
