// @progress/kendo-angular-layout/dist/es/common/draggable.directive.js
import * as tslib_1 from "tslib";
import { EventEmitter, ElementRef, Directive } from '@angular/core';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
var DraggableDirective = /** @class */ (function () {
    function DraggableDirective(element) {
        var _this = this;
        this.kendo = {
            drag: new EventEmitter(),
            press: new EventEmitter(),
            release: new EventEmitter()
        };
        if (typeof document !== 'undefined') {
            this.draggable = new Draggable({
                drag: function (e) { return _this.kendo.drag.next(e); },
                press: function (e) { return _this.kendo.press.next(e); },
                release: function (e) { return _this.kendo.release.next(e); }
            });
            this.draggable.bindTo(element.nativeElement);
        }
    }
    DraggableDirective.prototype.ngOnDestroy = function () {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    };
    DraggableDirective = tslib_1.__decorate([
        Directive({
            outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
            selector: '[kendoLayoutDraggable]'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], DraggableDirective);
    return DraggableDirective;
}());
export { DraggableDirective };
