// @progress/kendo-angular-dateinputs/dist/es/calendar/localization/calendar-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
var CalendarCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CalendarCustomMessagesComponent, _super);
    function CalendarCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    CalendarCustomMessagesComponent_1 = CalendarCustomMessagesComponent;
    Object.defineProperty(CalendarCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    CalendarCustomMessagesComponent = CalendarCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: Messages,
                    useExisting: forwardRef(function () { return CalendarCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                }
            ],
            selector: 'kendo-calendar-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], CalendarCustomMessagesComponent);
    return CalendarCustomMessagesComponent;
    var CalendarCustomMessagesComponent_1;
}(Messages));
export { CalendarCustomMessagesComponent };
