// @progress/kendo-angular-treeview/dist/es2015/checkbox/checkbox.component.js
import { Component, ElementRef, EventEmitter, Input, Output, HostBinding, Renderer2 } from '@angular/core';
import { guid } from '../utils';
/**
 * @hidden
 *
 * Represents the CheckBox component of the Kendo UI TreeView for Angular.
 *
 */
export class CheckBoxComponent {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        /**
         * Specifies the [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) of the component.
         */
        this.id = `_${guid()}`;
        /**
         * Specifies the checked state of the component.
         *
         * The available options are:
         * - `none`&mdash;Is not checked.
         * - `indeterminate`&mdash;Sets the indeterminate state.
         * - `checked`&mdash;Sets the checked state.
         */
        this.checkState = 'none';
        /**
         * Specifies the [`tabindex`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex) of the component.
         */
        this.tabindex = 0;
        /**
         * Fires when the user changes the check state of the component.
         */
        this.checkStateChange = new EventEmitter();
    }
    //XXX: implement ComponentValueAccessor
    //XXX: focus/blur methods
    get classWrapper() { return true; }
    get indeterminate() {
        return this.checkState === 'indeterminate';
    }
    get checked() {
        return this.checkState === 'checked';
    }
    ngOnInit() {
        this.renderer.removeAttribute(this.element.nativeElement, "tabindex");
    }
    handleChange(e) {
        this.checkStateChange.emit(e.target.checked ? 'checked' : 'none');
    }
}
CheckBoxComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-checkbox',
                template: `
        <input
            class="k-checkbox"
            type="checkbox"
            [id]="id"
            [checked]="checked"
            [indeterminate]="indeterminate"
            [tabindex]="tabindex"
            (change)="handleChange($event)"
        />
        <label
            class="k-checkbox-label"
            tabindex="-1"
            [for]="id"
        >{{labelText}}</label>
    `
            },] },
];
/** @nocollapse */
CheckBoxComponent.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
];
CheckBoxComponent.propDecorators = {
    'classWrapper': [{ type: HostBinding, args: ['class.k-checkbox-wrapper',] },],
    'id': [{ type: Input },],
    'checkState': [{ type: Input },],
    'labelText': [{ type: Input },],
    'tabindex': [{ type: Input },],
    'checkStateChange': [{ type: Output },],
};
