// @progress/kendo-angular-dropdowns/dist/es2015/list.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ViewChildren, QueryList, ElementRef, ViewChild } from '@angular/core';
import { ListItemDirective } from './list-item.directive';
import { ItemTemplateDirective } from './templates/item-template.directive';
import { isPresent, getter } from './util';
import { SelectionService } from './selection.service';
import { map } from 'rxjs/operators/map';
import { merge } from 'rxjs/operators/merge';
/**
 * @hidden
 */
let ListComponent = class ListComponent {
    constructor(selectionService) {
        this.data = [];
        this.selected = [];
        this.focused = -1;
        this.multipleSelection = false;
        this.onClick = new EventEmitter();
        this.selectionService = selectionService;
        this.scrollSubscription = this.selectionService
            .onSelect.pipe(map((args) => args.indices[0]), merge(this.selectionService.onFocus))
            .subscribe(this.scrollToItem.bind(this));
    }
    ngAfterViewInit() {
        if (this.show === true) {
            this.scrollToItem(this.selectionService.focused);
        }
    }
    ngOnDestroy() {
        this.scrollSubscription.unsubscribe();
    }
    setContainerClasses() {
        return {
            'k-list-scroller': true
        };
    }
    getHeight() {
        return this.height + "px";
    }
    getText(dataItem) {
        return getter(dataItem, this.textField);
    }
    getValue(dataItem) {
        return getter(dataItem, this.valueField);
    }
    scrollToItem(index) {
        const items = this.items.toArray();
        if (isPresent(items[index]) && index !== -1) {
            this.scroll(items[index].element);
        }
    }
    scroll(item) {
        if (!item) {
            return;
        }
        const nativeElement = item.nativeElement;
        let content = this.content.nativeElement, itemOffsetTop = nativeElement.offsetTop, itemOffsetHeight = nativeElement.offsetHeight, contentScrollTop = content.scrollTop, contentOffsetHeight = content.clientHeight, bottomDistance = itemOffsetTop + itemOffsetHeight;
        if (contentScrollTop > itemOffsetTop) {
            contentScrollTop = itemOffsetTop;
        }
        else if (bottomDistance > (contentScrollTop + contentOffsetHeight)) {
            contentScrollTop = (bottomDistance - contentOffsetHeight);
        }
        content.scrollTop = contentScrollTop;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], ListComponent.prototype, "data", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], ListComponent.prototype, "selected", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], ListComponent.prototype, "focused", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ListComponent.prototype, "textField", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ListComponent.prototype, "valueField", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], ListComponent.prototype, "height", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", ItemTemplateDirective)
], ListComponent.prototype, "template", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], ListComponent.prototype, "show", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ListComponent.prototype, "id", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ListComponent.prototype, "optionPrefix", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], ListComponent.prototype, "multipleSelection", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], ListComponent.prototype, "onClick", void 0);
tslib_1.__decorate([
    ViewChildren(ListItemDirective),
    tslib_1.__metadata("design:type", QueryList)
], ListComponent.prototype, "items", void 0);
tslib_1.__decorate([
    ViewChild('content'),
    tslib_1.__metadata("design:type", ElementRef)
], ListComponent.prototype, "content", void 0);
ListComponent = tslib_1.__decorate([
    Component({
        selector: 'kendo-list',
        template: `
    <div #content [ngClass]="setContainerClasses()" [style.maxHeight]="getHeight()" unselectable="on">
    <ul role="listbox" [attr.id]="id" [attr.aria-hidden]="!show" [ngClass]="{ 'k-list': true, 'k-reset': true }">
        <li role="option"
            *ngFor="let dataItem of data; let index = index"
            kendoDropDownsSelectable
            [index]="index"
            [multipleSelection]="multipleSelection"
            [attr.id]="optionPrefix + '-' + getValue(dataItem)"
            [attr.tabIndex]="-1"
            [ngClass]="{'k-item': true}">
            <ng-template *ngIf="template"
                [templateContext]="{
                    templateRef: template.templateRef,
                    $implicit: dataItem
                }">
            </ng-template>
            <ng-template [ngIf]="!template">{{ getText(dataItem) }}</ng-template>
        </li>
    </ul>
    </div>
  `
    }),
    tslib_1.__metadata("design:paramtypes", [SelectionService])
], ListComponent);
export { ListComponent };
