// @progress/kendo-angular-dropdowns/dist/es2015/selectable.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, HostBinding, HostListener } from '@angular/core';
import { SelectionService } from './selection.service';
/**
 * @hidden
 */
let SelectableDirective = class SelectableDirective {
    constructor(selectionService) {
        this.multipleSelection = false;
        this.selectionService = selectionService;
    }
    get focusedClassName() {
        return this.selectionService.isFocused(this.index);
    }
    get selectedClassName() {
        return this.selectionService.isSelected(this.index);
    }
    get ariaSelected() {
        return this.selectionService.isSelected(this.index) ? true : undefined;
    }
    onClick(event) {
        event.stopPropagation();
        if (this.multipleSelection) {
            if (this.selectionService.isSelected(this.index)) {
                this.selectionService.unselect(this.index);
            }
            else {
                this.selectionService.add(this.index);
            }
        }
        else {
            this.selectionService.change(this.index);
        }
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], SelectableDirective.prototype, "index", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SelectableDirective.prototype, "multipleSelection", void 0);
tslib_1.__decorate([
    HostBinding('class.k-state-focused'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SelectableDirective.prototype, "focusedClassName", null);
tslib_1.__decorate([
    HostBinding('class.k-state-selected'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SelectableDirective.prototype, "selectedClassName", null);
tslib_1.__decorate([
    HostBinding('attr.aria-selected'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], SelectableDirective.prototype, "ariaSelected", null);
tslib_1.__decorate([
    HostListener('click', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], SelectableDirective.prototype, "onClick", null);
SelectableDirective = tslib_1.__decorate([
    Directive({
        selector: '[kendoDropDownsSelectable]'
    }),
    tslib_1.__metadata("design:paramtypes", [SelectionService])
], SelectableDirective);
export { SelectableDirective };
