// @progress/kendo-angular-dateinputs/dist/es2015/datepicker/localization/datepicker-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
let DatePickerCustomMessagesComponent = DatePickerCustomMessagesComponent_1 = class DatePickerCustomMessagesComponent extends Messages {
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
};
DatePickerCustomMessagesComponent = DatePickerCustomMessagesComponent_1 = tslib_1.__decorate([
    Component({
        providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => DatePickerCustomMessagesComponent_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: 'kendo-datepicker-messages',
        template: ``
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], DatePickerCustomMessagesComponent);
export { DatePickerCustomMessagesComponent };
var DatePickerCustomMessagesComponent_1;
