// @progress/kendo-angular-gauges/dist/es/linear-gauge/linear-gauge.component.js
import * as tslib_1 from "tslib";
import { Component, NgZone, Input, ChangeDetectionStrategy, Optional, Inject } from '@angular/core';
import { ConfigurationService, ThemeService } from '../services';
import { IntlService } from '@progress/kendo-angular-intl';
import { RTL } from '@progress/kendo-angular-l10n';
import { GaugeComponent } from '../base-components';
import { LinearGauge } from '@progress/kendo-charts';
/**
 * The root Linear Gauge component.
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *   selector: 'my-app',
 *   template: `
 *      <kendo-lineargauge [pointer]="{ value: value }">
 *      </kendo-lineargauge>
 *   `
 * })
 * class AppComponent {
 *   public value: number = 10;
 * }
 *
 * ```
 */
var LinearGaugeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LinearGaugeComponent, _super);
    function LinearGaugeComponent(configurationService, themeService, intlService, ngZone, rtl) {
        return _super.call(this, configurationService, themeService, intlService, ngZone, rtl) || this;
    }
    LinearGaugeComponent.prototype.createInstance = function (element, options, theme, context) {
        this.instance = new LinearGauge(element, options, theme, context);
    };
    LinearGaugeComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    exportAs: 'kendoLinearGauge',
                    providers: [ConfigurationService],
                    selector: 'kendo-lineargauge',
                    template: "\n    <div #surface class='k-chart-surface'></div>\n    <kendo-resize-sensor (resize)=\"onResize($event)\" [rateLimit]=\"resizeRateLimit\"></kendo-resize-sensor>\n  "
                },] },
    ];
    /** @nocollapse */
    LinearGaugeComponent.ctorParameters = function () { return [
        { type: ConfigurationService, },
        { type: ThemeService, },
        { type: IntlService, },
        { type: NgZone, },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [RTL,] },] },
    ]; };
    LinearGaugeComponent.propDecorators = {
        'pointer': [{ type: Input },],
        'scale': [{ type: Input },],
    };
    return LinearGaugeComponent;
}(GaugeComponent));
export { LinearGaugeComponent };
