// @progress/kendo-angular-layout/dist/es2015/common/draggable.directive.js
import * as tslib_1 from "tslib";
import { EventEmitter, ElementRef, Directive } from '@angular/core';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
let DraggableDirective = class DraggableDirective {
    constructor(element) {
        this.kendo = {
            drag: new EventEmitter(),
            press: new EventEmitter(),
            release: new EventEmitter()
        };
        if (typeof document !== 'undefined') {
            this.draggable = new Draggable({
                drag: (e) => this.kendo.drag.next(e),
                press: (e) => this.kendo.press.next(e),
                release: (e) => this.kendo.release.next(e)
            });
            this.draggable.bindTo(element.nativeElement);
        }
    }
    ngOnDestroy() {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    }
};
DraggableDirective = tslib_1.__decorate([
    Directive({
        outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
        selector: '[kendoLayoutDraggable]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef])
], DraggableDirective);
export { DraggableDirective };
