// @progress/kendo-angular-layout/dist/es2015/tabstrip/tabstrip-tab.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, Input, ContentChildren, QueryList } from '@angular/core';
import { TabContentDirective } from './tab-content.directive';
import { TabTitleDirective } from './tab-title.directive';
/**
 * Represents the Kendo UI TabStrip Tab component for Angular.
 */
let TabStripTabComponent = class TabStripTabComponent {
    /**
     * Represents the Kendo UI TabStrip Tab component for Angular.
     */
    constructor() {
        this.active = false;
        this._tabContent = new QueryList();
    }
    get tabContent() {
        return this._tabContent.first;
    }
    ngAfterContentInit() {
        this.active = this.selected;
    }
    ngOnChanges(changes) {
        if (changes['selected'] && !changes['selected'].isFirstChange()) {
            this.active = this.selected;
        }
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], TabStripTabComponent.prototype, "title", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], TabStripTabComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], TabStripTabComponent.prototype, "selected", void 0);
tslib_1.__decorate([
    ContentChildren(TabContentDirective),
    tslib_1.__metadata("design:type", QueryList)
], TabStripTabComponent.prototype, "_tabContent", void 0);
tslib_1.__decorate([
    ContentChild(TabTitleDirective),
    tslib_1.__metadata("design:type", TabTitleDirective)
], TabStripTabComponent.prototype, "tabTitle", void 0);
TabStripTabComponent = tslib_1.__decorate([
    Component({
        exportAs: 'kendoTabStripTab',
        selector: 'kendo-tabstrip-tab',
        template: ``
    })
], TabStripTabComponent);
export { TabStripTabComponent };
