// @progress/kendo-angular-gauges/dist/es2015/radial-gauge/radial-gauge.component.js
import { Component, NgZone, Input, ChangeDetectionStrategy, Optional, Inject } from '@angular/core';
import { ConfigurationService, ThemeService } from '../services';
import { IntlService } from '@progress/kendo-angular-intl';
import { RTL } from '@progress/kendo-angular-l10n';
import { GaugeComponent } from '../base-components';
import { RadialGauge } from '@progress/kendo-charts';
/**
 * The root Radial Gauge component.
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *   selector: 'my-app',
 *   template: `
 *      <kendo-radialgauge [pointer]="{ value: value }">
 *      </kendo-radialgauge>
 *   `
 * })
 * class AppComponent {
 *   public value: number = 10;
 * }
 *
 * ```
 */
export class RadialGaugeComponent extends GaugeComponent {
    constructor(configurationService, themeService, intlService, ngZone, rtl) {
        super(configurationService, themeService, intlService, ngZone, rtl);
    }
    createInstance(element, options, theme, context) {
        this.instance = new RadialGauge(element, options, theme, context);
    }
}
RadialGaugeComponent.decorators = [
    { type: Component, args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'kendoRadialGauge',
                providers: [ConfigurationService],
                selector: 'kendo-radialgauge',
                template: `
    <div #surface class='k-chart-surface'></div>
    <kendo-resize-sensor (resize)="onResize($event)" [rateLimit]="resizeRateLimit"></kendo-resize-sensor>
  `
            },] },
];
/** @nocollapse */
RadialGaugeComponent.ctorParameters = () => [
    { type: ConfigurationService, },
    { type: ThemeService, },
    { type: IntlService, },
    { type: NgZone, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [RTL,] },] },
];
RadialGaugeComponent.propDecorators = {
    'pointer': [{ type: Input },],
    'scale': [{ type: Input },],
};
