// @progress/kendo-angular-layout/dist/es/tabstrip/tabstrip-tab.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, Input, ContentChildren, QueryList } from '@angular/core';
import { TabContentDirective } from './tab-content.directive';
import { TabTitleDirective } from './tab-title.directive';
/**
 * Represents the Kendo UI TabStrip Tab component for Angular.
 */
var TabStripTabComponent = /** @class */ (function () {
    function TabStripTabComponent() {
        this.active = false;
        this._tabContent = new QueryList();
    }
    Object.defineProperty(TabStripTabComponent.prototype, "tabContent", {
        get: function () {
            return this._tabContent.first;
        },
        enumerable: true,
        configurable: true
    });
    TabStripTabComponent.prototype.ngAfterContentInit = function () {
        this.active = this.selected;
    };
    TabStripTabComponent.prototype.ngOnChanges = function (changes) {
        if (changes['selected'] && !changes['selected'].isFirstChange()) {
            this.active = this.selected;
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], TabStripTabComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], TabStripTabComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], TabStripTabComponent.prototype, "selected", void 0);
    tslib_1.__decorate([
        ContentChildren(TabContentDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], TabStripTabComponent.prototype, "_tabContent", void 0);
    tslib_1.__decorate([
        ContentChild(TabTitleDirective),
        tslib_1.__metadata("design:type", TabTitleDirective)
    ], TabStripTabComponent.prototype, "tabTitle", void 0);
    TabStripTabComponent = tslib_1.__decorate([
        Component({
            exportAs: 'kendoTabStripTab',
            selector: 'kendo-tabstrip-tab',
            template: ""
        })
    ], TabStripTabComponent);
    return TabStripTabComponent;
}());
export { TabStripTabComponent };
