// @progress/kendo-angular-dateinputs/dist/es2015/timepicker/localization/timepicker-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
let TimePickerCustomMessagesComponent = TimePickerCustomMessagesComponent_1 = class TimePickerCustomMessagesComponent extends Messages {
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
};
TimePickerCustomMessagesComponent = TimePickerCustomMessagesComponent_1 = tslib_1.__decorate([
    Component({
        providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => TimePickerCustomMessagesComponent_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: 'kendo-timepicker-messages',
        template: ``
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], TimePickerCustomMessagesComponent);
export { TimePickerCustomMessagesComponent };
var TimePickerCustomMessagesComponent_1;
