// @progress/kendo-angular-dateinputs/dist/es2015/timepicker/localization/timeselector-localized-messages.directive.js
import * as tslib_1 from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * @hidden
 */
let TimeSelectorLocalizedMessagesDirective = TimeSelectorLocalizedMessagesDirective_1 = class TimeSelectorLocalizedMessagesDirective extends Messages {
    constructor(service) {
        super();
        this.service = service;
    }
};
TimeSelectorLocalizedMessagesDirective = TimeSelectorLocalizedMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => TimeSelectorLocalizedMessagesDirective_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: '[kendoTimeSelectorLocalizedMessages]'
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], TimeSelectorLocalizedMessagesDirective);
export { TimeSelectorLocalizedMessagesDirective };
var TimeSelectorLocalizedMessagesDirective_1;
