// @progress/kendo-angular-dateinputs/dist/es2015/timepicker/localization/timeselector-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * @hidden
 *
 * Custom component messages override default component messages.
 */
let TimeSelectorCustomMessagesComponent = TimeSelectorCustomMessagesComponent_1 = class TimeSelectorCustomMessagesComponent extends Messages {
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
};
TimeSelectorCustomMessagesComponent = TimeSelectorCustomMessagesComponent_1 = tslib_1.__decorate([
    Component({
        providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => TimeSelectorCustomMessagesComponent_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: 'kendo-timeselector-messages',
        template: ``
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], TimeSelectorCustomMessagesComponent);
export { TimeSelectorCustomMessagesComponent };
var TimeSelectorCustomMessagesComponent_1;
