// @progress/kendo-angular-dateinputs/dist/es/dateinput/localization/dateinput-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
var DateInputCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DateInputCustomMessagesComponent, _super);
    function DateInputCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    DateInputCustomMessagesComponent_1 = DateInputCustomMessagesComponent;
    Object.defineProperty(DateInputCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    DateInputCustomMessagesComponent = DateInputCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: Messages,
                    useExisting: forwardRef(function () { return DateInputCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                }
            ],
            selector: 'kendo-dateinput-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], DateInputCustomMessagesComponent);
    return DateInputCustomMessagesComponent;
    var DateInputCustomMessagesComponent_1;
}(Messages));
export { DateInputCustomMessagesComponent };
