// @progress/kendo-angular-dateinputs/dist/es2015/calendar/localization/calendar-custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
let CalendarCustomMessagesComponent = CalendarCustomMessagesComponent_1 = class CalendarCustomMessagesComponent extends Messages {
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
};
CalendarCustomMessagesComponent = CalendarCustomMessagesComponent_1 = tslib_1.__decorate([
    Component({
        providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => CalendarCustomMessagesComponent_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: 'kendo-calendar-messages',
        template: ``
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], CalendarCustomMessagesComponent);
export { CalendarCustomMessagesComponent };
var CalendarCustomMessagesComponent_1;
